aws/retry/middleware.go

```diff
@@ -87,10 +90,19 @@ func (r Attempt) HandleFinalize(ctx context.Context, in smithymiddle.FinalizeInp
 			AttemptClockSkew: attemptClockSkew,
 		})
 
+		var attemptMetadata requestAttempt
+		defer func() {
+			allAttemptMetadata.Attempts = append(allAttemptMetadata.Attempts, attemptMetadata)
+		}()
+
```

Comment:
defer's inside of for loops are generally not a good idea as they can lead to an unexpected behavior when they are unrolled.  If using defer here the content of the `for` probably should be moved to a function so that there are no unexpected behaviors with the defer.